File: Hw7 (Java)/src/HW7_171044023/BoardArray2D.java
package HW7_171044023;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class BoardArray2D extends AbstractBoard{
	
	/**
	 * This array of string is the 2D game board.
	 */
	private String [][] board;
	
	/**
	 * BoardArray2D no-argument constructor
	 */
	public BoardArray2D() {
		super();
	}
	
	/**
	 * BoardArray2D two-arguments constructor
	 * @param newRow int value of row number
	 * @param newColumn int value of column number
	 */
	public BoardArray2D(int newRow, int newColumn) {
		super(newRow, newColumn);	
		board = new String[row][column];
	}
	
	/**
	 * This method returns string representation of BoardArray2D object.
	 *@return string of representation of BoardArray2D object
	 */
	@Override
	public String toString() {
		String res = board[0][0];
		
		for(int i=0; i<getRow(); i++) {
			for(int j=0; j<getColumn(); j++) {
				if(j == 0 && i == 0)	res += " "; 
				else{
					res += board[i][j];
					res += " ";
				}
			}
			res += "\n";
		}
		res += "\n";
		
		return res;
	}

	/**
	 * This method reads the board from the file given as parameter.
	 * @param file_name string of the file name
	 */
	@Override
	public void readFromFile(String file_name) {
		findSize(file_name);
		
		try {
		    FileReader read = new FileReader(file_name);
	        BufferedReader buffer = new BufferedReader(read);

	        String line;
	       
	        int empty = 0, begin = 0;
	        int i = 0; //line number (row)
	        
	        while ((line = buffer.readLine()) != null) {
	        	for(int j=0; j<getColumn(); j++) {
		        	// find first empty
		        	if(j == 0)	empty = line.indexOf(" ");
		        	
		        	//board == line
		        	if(empty == -1)	board[i][j] = line.substring(begin, line.length());
		        	else	board[i][j] = line.substring(begin, empty);

		        	begin = empty+1;
			        //find next empty
		        	empty = line.indexOf(" ", empty+1);
	        	}
	        	i++;
            }
			buffer.close();
			findEmptyCell();
			calculateWallNumber();
			fillCheckBoard();
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }

	}

	/**
	 * This method writes the board to the file given as parameter.
	 * @param file_name string of the file name
	 */
	@Override
	public void writeToFile(String file_name) {
		try {
			FileWriter write = new FileWriter(file_name);
			BufferedWriter buffer = new BufferedWriter(write);
			
			for(int i=0; i<getRow(); i++) {
				for(int j=0; j<getColumn(); j++) {
					buffer.write(board[i][j]);
					
					if(j != getColumn()-1 ) 
						buffer.write(" ");
				}
				buffer.newLine();
			}
			
			buffer.close();
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }

	}

	/**
	 *This method resets the board to the solution.
	 */
	@Override
	public void reset() {
		int k = 0;
	
		for(int i=0; i<getRow(); i++){ // I filled the board with the numbers in the array(numbers).
			for(int j=0; j<getColumn(); j++){
				if(board[i][j].compareTo("00") != 0){
					board[i][j] = checkBoard[k];
					k++;
				}
			}
		}
		setNumberOfMoves(0);
		
	}

	/**
	 * This method sets the board size to given values.
	 * The values are given as parameters and there are no restrictions on the board size.
	 * The board is reset after this operation.
	 * @param size1 int value of board's first size
	 * @param size2 int value of board's second size
	 */
	@Override
	public void setSize(int size1, int size2) {
		setRow(size1);
		setColumn(size2);
		calculateWallNumber();
		fillCheckBoard();
		reset();
	}

	/**
	 * This method makes a move according to the given char parameter.
	 * @param choice int value of user's choice for the move
	 */
	@Override
	public void move(char choice) {
		boolean flag = false;
		findEmptyCell();
		switch(choice){
			case 'L':	case 'l':	//LEFT
				if(getEmptyY() > 0)
					flag = movement(empty_x, empty_y-1);
				else	System.out.println("LEFT is invalid move!");
			break;

			case 'R':	case 'r':	//RIGHT
				if(getEmptyY() < getColumn()-1)
					flag = movement(empty_x, empty_y+1);
				else	System.out.println("RIGHT is invalid move!");
			break;

			case 'U':	case 'u':	//UP
				if(getEmptyX() > 0)
					flag = movement(empty_x-1, empty_y);
				else	System.out.println("UP is invalid move!");
			break;

			case 'D':	case 'd':	//DOWN
				if(getEmptyX() < getRow()-1)
					flag = movement(empty_x+1, empty_y);
				else	System.out.println("DOWN is invalid move!");
			break;
		}
		if(flag == true){
			findEmptyCell();
			setNumberOfMoves( numberOfMoves() + 1 );
			setLastMove(choice);
		}
	}

	/**
	 * This method checks if the board is solved..
	 * @return returns true if the board is a solution
	 */
	@Override
	public boolean isSolved() {
		int k = 0;
		System.out.println();
		for(int i=0; i<getRow(); i++){
			for(int j=0; j<getColumn(); j++){
				if(board[i][j].compareTo("00") != 0){
					if( board[i][j].compareTo(checkBoard[k]) != 0)	return false;
					k++;
				}
			}
		}
		return true;
	}

	/**
	 * This method takes two indexes and returns the corresponding cell content.
	 * @param x int value of first index
	 * @param y int value of second index
	 * @return returns the value in the given coordinates.
	 */
	@Override
	public String cell(int x, int y) {
		return board[x][y];
	}

	/**
	 * This method controls equality.
	 * Two boards are equal, if the boards are the same.
	 * This method does not consider last move or the number of steps.
	 * @param other the object of AbstractBoard
	 * @return returns true if two boards are equal.
	 */
	@Override
	public boolean Equals(AbstractBoard other) {
		
		if(getRow() == other.getRow() && getColumn() ==  other.getColumn()) {
			for(int i=0; i<getRow(); i++) {
				for(int j=0; j<getColumn(); j++) {
					if( board[i][j].compareTo( other.cell(i, j) ) != 0)	return false;
				}
			}
		}
		else	return false;
		
		return true;
	}
	
	
	/**
	 * This method swaps the empty cell and the cell in the index given as parameter.
	 * @param index int value of the index of the cell to move
	 * @return returning true, if the cell can move.
	 */
	private boolean movement(int x, int y){
		String temp;
		
		findEmptyCell(); // this function find the empty cell's current coordinates.
		if(board[x][y].compareTo("00") != 0){ // if the block is wall, the empty cell do not move.
			temp = board[empty_x][empty_y];
			board[empty_x][empty_y] = board[x][y];
			board[x][y] = temp;
			return true;
		}
		else	System.out.println("Empty cell cannot move!");
		return false;
	}

	/**
	 * This method calculates the number of wall in the board.
	 */
	private void calculateWallNumber() {
		int count = 0;
		
		for(int i=0; i<getRow(); i++) {
			for(int j=0; j<getColumn(); j++) {
				if(board[i][j].compareTo("00") == 0) count++;
			}
		}
		
		setWallNumber(count);
	}
		
	/**
	 * This method finds the empty cell's current coordinates.
	 */
	private void findEmptyCell() {
		int x = 0, y = 0;
			
		for(int i=0; i<getRow(); i++) {
			for(int j=0; j<getColumn(); j++) {
				if(board[i][j].compareTo("bb") == 0) {
					x = i;
					y = j;
				}
			}
		}
			
		setEmptyCell(x, y);
	}
	
	/**
	 * This method finds the sizes of the board read from the file.
	 * @param file_name the string of the name of the file to be read
	 */
	private void findSize(String file_name) {
		try {
		    FileReader read = new FileReader(file_name);
	        BufferedReader buffer = new BufferedReader(read);
	        
	        int size1 = 0;
	        int size2 = 1;
	        
	        String line;
	        
	        while ((line = buffer.readLine()) != null) {
	        	if(size1 == 0){
	        		int i = 0;
		        	while(i < line.length() ) {
		        		if(line.charAt(i) == ' ') size2++;
		        		i++;
		        	}
	        	}
	        	size1++;
            }
			buffer.close();
			
			setRow(size1);
			setColumn(size2);
			
			board = new String[getRow()][getColumn()];	
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }
	}
}

File: Hw5 (C++)/boardarray2d.cpp
#include<iostream>
#include <string>
#include<fstream>
#include<cstdlib>
#include<time.h>
using namespace std;
#include "boardarray2d.h"

namespace Boards{
	BoardArray2D::BoardArray2D() : AbstractBoard(){
		finalBoard = new string[0];
	}

	BoardArray2D::BoardArray2D(const BoardArray2D& other) : AbstractBoard(other){

		finalBoard = new string[row*column-wall_number];
		for(int i=0; i<row*column-wall_number; i++)
			finalBoard[i] = other.finalBoard[i];

		board = new string*[row];
		for(int i=0; i<row; i++)
			board[i] = new string[column];

		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++)
				board[i][j] = other.board[i][j];
		}
	}

	void BoardArray2D::operator=(const BoardArray2D& other){
		empty_x = other.empty_x;
		empty_y = other.empty_y;

		if(row != other.row && column != other.column){
			delete [] finalBoard;
			finalBoard = new string[other.row*other.column-other.wall_number];

			for(int i=0; i<row; i++)
				delete [] board[i];
			delete [] board;
			board = new string*[other.row];
			for(int i=0; i<other.row; i++)
				board[i] = new string[other.column];
		}
		row = other.row;
		column = other.column;
		wall_number = other.wall_number;

		for(int i=0; i<row*column-wall_number; i++)
			finalBoard[i] = other.finalBoard[i];	

		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++)
				board[i][j] = other.board[i][j];
		}
	}

	string BoardArray2D::operator()(const int x, const int y) const{
		return board[x][y];
	}

	bool BoardArray2D::operator==(AbstractBoard& other){
		if(this->row == other.getRow() && this->column == other.getColumn()){
			for(int i=0; i<row; i++){
				for(int j=0; j<column; j++){
					if(this->operator()(i,j) != other(i,j))	return false;
				}
			}
		}
		else	return false;
		return true;
	}

	void BoardArray2D::setSize(int size1, int size2){
		row = size1;
		column = size2;
		board = new string*[row];
		for(int i=0; i<row; i++)
			board[i] = new string[column];
		setFinalBoard();
	}

	void BoardArray2D::print() const{
		int i, j;
		cout << endl;
		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				cout.width(5);
				if(board[i][j] == "bb")	cout << " ";
				else if(board[i][j] == "00")	cout << "*";
				else	cout << board[i][j];
			}
			cout << endl;
		}
		cout << endl;
	}

	void BoardArray2D::calculateWallNumber(){ // this function calculates the number of wall in the game.
		int i, j;
		wall_number = 0;

		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				if(board[i][j] == "00") wall_number++;
			}
		}
	}

	void BoardArray2D::setFinalBoard(){ //this function fills the array accordig to the final board.
		calculateWallNumber();
		finalBoard = new string[row*column-wall_number];

		if(row*column-wall_number > 0){
			int i, j, num1 = 50, num2 = 49;
			string number = "01";

			for(i=0; i<row*column-1-wall_number; i++){
			// I filled the array with the numbers in the game with using ascii table. I used the array to compare the correct board with the current board and fill the final board. 
				finalBoard[i] = number;
				if(num1 > 57){
					num1 = 48;
					number[0] = num2;
					num2++;
				}
				number[1] = num1;
				num1++;
			}

			finalBoard[row*column-1-wall_number] = "bb"; // I put 0 on the last element of the array for the empty cell.
		}
	}

	void BoardArray2D::initRandomBoard(){
		srand(time(NULL));
		int temp[9][9], control_array[row*column], i, j, counter;
		bool flag;
		auto size = 0; // this variable is the size of control_array

		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				do{
					counter = 0;
					temp[i][j] = rand() % (row*column);
					for(int n=0; n<size; n++)
						if(temp[i][j] == control_array[n]) counter++; // if the random number exists in the array, this counter increases. so flag is false.
					if(counter == 0)	flag = true;
					else	flag = false;
				}while(flag != true);
				control_array[size] = temp[i][j];
				size++;
			}
		}

		//fill the board
		string number[row*column];
		int k = 0;
		//I convert integer array to string array.
		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				if(temp[i][j] > 9){
					number[k] += (temp[i][j] / 10 + 48);
					number[k] += (temp[i][j] % 10 + 48);
				}
				else{
					number[k] += '0';
					number[k] += (temp[i][j] + 48);
				}
				if(number[k] == "00")	board[i][j] = "bb";
				else	board[i][j] = number[k];
				k++;
			}
		}
		findEmptyCell();
	}

	bool BoardArray2D::isSolved(){
		//this function checks if this board is solved.
		int i, j, num1 = 50, num2 = 49;
		string number = "01";

		int k = 0;
		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				if(board[i][j] != "00"){
					if( board[i][j] !=  finalBoard[k] )	return false;
					k++;
				}
			}
		}
		return true;
	}

	void BoardArray2D::reset(){
		int i, j;

		auto k = 0;
		for(i=0; i<row; i++){ // I filled the board with the numbers in the array(numbers).
			for(j=0; j<column; j++){
				if(board[i][j] != "00"){
					board[i][j] = finalBoard[k];
					k++;
				}
			}
		}
		number_of_moves = 0;
	}

	void BoardArray2D::readFromFile(string file_name){ // this function reads the file.
		setSize(0, 0);
		string line;
		bool flag;
		int size1 = 0, size2 = 0;

		fstream gameBoard;
		do{
			gameBoard.open(file_name);
			if( gameBoard.is_open() ){
			flag = true;
			}
			else{
				cout << "Please enter the file name(.txt) you want to play." << endl;
				cin >> file_name;
				flag = false;
			}
		}while(flag != true);

		while( getline(gameBoard, line) )
			size1++; // the number of row
		
		int i = 0;
		while(i < 25){
				if(line[i] == ' ')	size2++; // the number of column
				i++;
		}
		size2 += 1;
		gameBoard.close();

		setSize(size1, size2);
		int x = 0, y = 0;
		gameBoard.open(file_name);
		while(gameBoard >> board[x][y]){
			y++;
			if(y % size2 == 0){ // when reaches the end of the line, the number of rows increases.
				y = 0;
				x++;
			}
		}
		gameBoard.close();

		findEmptyCell();
		setFinalBoard(); /* I call this function again(previously called in setSize(line238))
		because the wall_number was not known because the file was not read.*/
		setNumberOfMoves(0);
	}

	void BoardArray2D::writeToFile() const{
		string file_name;
		int i, j;

		cout << "Please enter the file name(.txt) you want to save: " << endl;
		cin >> file_name;
		
		ofstream saveBoard;
		saveBoard.open(file_name);
		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				if(j == column-1)	saveBoard << board[i][j];
				else	saveBoard << board[i][j] << " ";
			}
			saveBoard << endl;
		}
		saveBoard.close();
	}

	void BoardArray2D::move(char choice){
		bool flag;
		findEmptyCell();
		switch(choice){
			case 'L':	case 'l':	//LEFT
				if(empty_y > 0)
					flag = movement(empty_x, empty_y-1);
			break;

			case 'R':	case 'r':	//RIGHT
				if(empty_y < column-1)
					flag = movement(empty_x, empty_y+1);
			break;

			case 'U':	case 'u':	//UP
				if(empty_x > 0)
					flag = movement(empty_x-1, empty_y);
			break;

			case 'D':	case 'd':	//DOWN
				if(empty_x < row-1)
					flag = movement(empty_x+1, empty_y);
			break;
		}
		if(flag == true){
			if(choice == 'L' || choice == 'l')	cout << "your move is Left" << endl;
			else if(choice == 'R' || choice == 'r')	cout << "your move is Right" << endl;
			else if(choice == 'U' || choice == 'u')	cout << "your move is Up" << endl;
			else if(choice == 'D' || choice == 'd')	cout << "your move is Down" << endl;
			number_of_moves ++;
			setLastMove(choice);
		}
	}

	bool BoardArray2D::movement(int x, int y){ // this function swaps the two coordinates.
		string temp;

		findEmptyCell(); // this function find the empty cell's current coordinates.
		if(board[x][y] != "00"){ // if the block is wall, the empty cell do not move.
			temp = board[empty_x][empty_y];
			board[empty_x][empty_y] = board[x][y];
			board[x][y] = temp;
			return true;
		}
		return false;
	}

	void BoardArray2D::findEmptyCell(){ // this function find the empty cell's current coordinates.
		int i, j;
		for(i=0; i<row; i++){
			for(j=0; j<column; j++){
				if(board[i][j] == "bb"){
					empty_x = i;
					empty_y = j;
				}
			}
		}
	}

	BoardArray2D::~BoardArray2D(){
		delete [] finalBoard;

		for(int i=0; i<row; i++)
			delete [] board[i];
		delete [] board;
	}
}